### PR TITLE
Update flex volume 1.6 documentation.

### PR DESCRIPTION
**Pull request by chakri-nelluri**
**What this PR does / why we need it**:
Update documentation for 1.6 flex volume changes.

### CODE REVIEW

**Comment by chakri-nelluri** [comment]
@saad-ali @rootfs PTAL

File: examples/volumes/flexvolume/README.md [hunk 1]
```diff
@@ -1,54 +1,94 @@
 # Flexvolume
 
-Flexvolume enables users to mount vendor volumes into kubernetes. It expects vendor drivers are installed in the volume plugin path on every kubelet node.
-
-It allows for vendors to develop their own drivers to mount volumes on nodes.
+Flexvolume enables users to write their own drivers and add support for their volumes in Kubernetes. Vendor drivers should be installed in the volume plugin path on every Kubelet node and on master node(s) if "-enable-controller-attach-detach" Kubelet option is enabled.
```
Comment (klausenbusk) [inline examples/volumes/flexvolume/README.md]: Missing hyphen (-) `--enable-controller-attach-detach`?

File: examples/volumes/flexvolume/README.md [hunk 2]
```diff
@@ -1,54 +1,94 @@
...
+```
+
+#### Wait for attach:
+Wait for the volume to be attached on the remote node. On success, the path to the device is returned.
+
+```
+<driver executable> waitfordetach <mount device> <json options>
+```
+
+#### Volume is Attached:
```
Comment (klausenbusk) [inline examples/volumes/flexvolume/README.md]: Only relevant if `--enable-controller-attach-detach` is set correct?

File: examples/volumes/flexvolume/README.md [hunk 3]
```diff
@@ -1,54 +1,94 @@
...
 
 ## Plugin details
-
-Driver will be invoked with 'Init' to initialize the driver. It will be invoked with 'attach' to attach the volume and with 'detach' to detach the volume from the kubelet node. It also supports custom mounts using 'mount' and 'unmount' callouts to the driver.
+The plugin expects the following call-outs are implemented for the backend drivers. Some call-outs are optional. Call-outs are invoked from the Kubelet & Controller manager nodes.
 
 ### Driver invocation model:
 
-Init:
+#### Init:
```
Comment (saad-ali) [inline examples/volumes/flexvolume/README.md]: When is this called by kubernetes? From what component (kubelet?)? What is it expected to do?

Comment (saad-ali) [inline examples/volumes/flexvolume/README.md]: These questions should be answered for each callout.

File: examples/volumes/flexvolume/README.md [hunk 4]
```diff
@@ -1,54 +1,94 @@
...
+Check the volume is attached on the node.
+
+```
+<driver executable> isattached <json options> <node name>
+```
+
+#### Mount device:
+Mount device mounts the device to a global path which individual pods can then bind mount.
+
+This call-out does not pass "secrets" specified in Flexvolume spec. If your driver requires secrets, do not implement this call-out and instead use "mount" call-out and implement attach and mount in that call-out.
```
Comment (saad-ali) [inline examples/volumes/flexvolume/README.md]: This is the mount call?

File: examples/volumes/flexvolume/README.md [hunk 5]
```diff
@@ -1,54 +1,94 @@
...
-Attach:
+#### Get volume name:
+Get a cluster wide unique volume name for the volume.
+
+```
+<driver executable> getvolumename <json options>
+```
+
+#### Attach:
+Attach the volume specified by the given spec on the given host. On success, returns the device path where the device is attached on the node. Nodename param is only valid if "-enable-controller-attach-detach" Kubelet option is enabled.
```
Comment (saad-ali) [inline examples/volumes/flexvolume/README.md]: >  Nodename param is only valid if "-enable-controller-attach-detach" Kubelet option is enabled.

When attach/detach controller is disabled, kubelet calls attach, and I believe still provides the node name, no?

**Review by saad-ali — CHANGES_REQUESTED** [review]

**Comment by chakri-nelluri** [comment]
Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions.

---

*[examples/volumes/flexvolume/README.md, line 3 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/43326#-KfT1UHmzB-euEalW4lW-r1-3:-KfgqYTant_KuWPryj3Z:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/522cf1876e3edc7b560e8e3b8d3c65cb6ebc95f9/examples/volumes/flexvolume/README.md#L3)):*
<details><summary><i>Previously, klausenbusk (Kristian Klausen) wrote…</i></summary><blockquote>

Missing hyphen (-) `--enable-controller-attach-detach`?
</blockquote></details>

Done.

---

*[examples/volumes/flexvolume/README.md, line 18 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/43326#-KfT1UHmzB-euEalW4lW-r1-18:-KfgqPimLutsCsXeHvTK:b-ex66us) ([raw file](https://github.com/kubernetes/kubernetes/blob/522cf1876e3edc7b560e8e3b8d3c65cb6ebc95f9/examples/volumes/flexvolume/README.md#L18)):*
<details><summary><i>Previously, saad-ali (Saad Ali) wrote…</i></summary><blockquote>

These questions should be answered for each callout.
</blockquote></details>

Added "Called from Kubelet & Controller manager" for relevant callouts

---

*[examples/volumes/flexvolume/README.md, line 33 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/43326#-KfT1UHmzB-euEalW4lW-r1-33:-KfglfLT-91rdRjJFTOc:bh6vzbw) ([raw file](https://github.com/kubernetes/kubernetes/blob/522cf1876e3edc7b560e8e3b8d3c65cb6ebc95f9/examples/volumes/flexvolume/README.md#L33)):*
<details><summary><i>Previously, saad-ali (Saad Ali) wrote…</i></summary><blockquote>

>  Nodename param is only valid if "-enable-controller-attach-detach" Kubelet option is enabled.

When attach/detach controller is disabled, kubelet calls attach, and I believe still provides the node name, no?
</blockquote></details>

Trying to signify it still provides but nodename but it not valid/relevant.

---

*[examples/volumes/flexvolume/README.md, line 54 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/43326#-KfT1UHmzB-euEalW4lW-r1-54:-Kfgmelt2LY-96Wm7MFG:bwmh4pf) ([raw file](https://github.com/kubernetes/kubernetes/blob/522cf1876e3edc7b560e8e3b8d3c65cb6ebc95f9/examples/volumes/flexvolume/README.md#L54)):*
<details><summary><i>Previously, klausenbusk (Kristian Klausen) wrote…</i></summary><blockquote>

Only relevant if `--enable-controller-attach-detach` is set correct?
</blockquote></details>

It is getting called from both Kubelet & Controller manager right.

---

*[examples/volumes/flexvolume/README.md, line 64 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/43326#-KfT1UHmzB-euEalW4lW-r1-64:-KfgntSJPtgnzN-vnqQI:b-rbd8cq) ([raw file](https://github.com/kubernetes/kubernetes/blob/522cf1876e3edc7b560e8e3b8d3c65cb6ebc95f9/examples/volumes/flexvolume/README.md#L64)):*
<details><summary><i>Previously, saad-ali (Saad Ali) wrote…</i></summary><blockquote>

This is the mount call?
</blockquote></details>

this is the mountdevice call-out.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/43326)*

**Comment by devin-donnelly** [comment]
@saad-ali Does this need to be documented on kubernetes.io (the web documentation)?

**Comment by saad-ali** [comment]
> @saad-ali Does this need to be documented on kubernetes.io (the web documentation)?

Good question. Intended audience for Flex volume plugin is Kubernetes volume plugin developers. So maybe not?

**Comment by saad-ali** [comment]
Other than the concern's @devin-donnelly raised about the correct place for this doc, LGTM

**Comment by chakri-nelluri** [comment]
Thanks @saad-ali 

@devin-donnelly I agree with @saad-ali. This is only intended for volume plugin drivers. I don't think we need to update documentation on kubernetes.io.

**Review by saad-ali — APPROVED** [review]

**Pull request merged**